lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+                                          FieldReferenceGraph::Node *To) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG,
+                  dbgs() << "Reconnect edge: (" << FromNode->Id << ","
+                         << ToNode->Id << ") to (" << From->Id << "," << To->Id
+                         << ") with " << DEBUG_PRINT_COUNT(ExecutionCount)
+                         << " and " << DEBUG_PRINT_DIST(DataSize) << "\n");
+  // Case #1
+  if (From == To) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_CPG, dbgs() << "Reconnection results in a node "
+                                              "pointing to itself, remove this "
+                                              "edge\n");
+    FromNode->OutEdges.erase(this);
+    ToNode->InEdges.erase(this);
+    FromNode = NULL;
+    ToNode = NULL;
+    return;
+  }
+  assert(From != FromNode || To != ToNode); // Assume never reconnect two nodes
+                                            // already connected with this Edge
+  Edge *ExistEdge = NULL;
+  for (auto *E : From->OutEdges) {
+    if (E->ToNode == To)
+      ExistEdge = E;
+  }
+  // Case #2
+  if (ExistEdge) {
+    assert(To->InEdges.find(ExistEdge) != To->InEdges.end());
+    DEBUG_WITH_TYPE(DEBUG_TYPE_CPG,
+                    dbgs() << "Node " << From->Id << " and Node " << To->Id
+                           << " already have an edge with "
+                           << DEBUG_PRINT_COUNT(ExistEdge->ExecutionCount)
+                           << " and " << DEBUG_PRINT_DIST(ExistEdge->DataSize)
+                           << "\n");
+    // Reconnect edge to two nodes that have connection, weights need to be
+    // adjusted
+    updatePairByMerging(ExecutionCount, DataSize, ExistEdge->ExecutionCount,
+                        ExistEdge->DataSize);
+    From->OutEdges.erase(ExistEdge);
+    To->InEdges.erase(ExistEdge);
+  }
+  // Case #3
+  if (From != FromNode)
+    FromNode->OutEdges.erase(this);
+  if (To != ToNode)
+    ToNode->InEdges.erase(this);
+  From->OutEdges.insert(this);
+  To->InEdges.insert(this);
+  FromNode = From;
+  ToNode = To;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG,
+                  dbgs() << "After reconnection: (" << FromNode->Id << ","
+                         << ToNode->Id << ") with "
+                         << DEBUG_PRINT_COUNT(ExecutionCount) << " and "
+                         << DEBUG_PRINT_DIST(DataSize) << "\n");
+}
+
+FieldReferenceGraph::BasicBlockHelperInfo *
+FieldReferenceGraph::createBasicBlockHelperInfo(const BasicBlock *BB) {
+  assert(BBInfoMap.find(BB) == BBInfoMap.end());
+  auto *ret = BBInfoMap[BB] = new BasicBlockHelperInfo;
+  return ret;
+}
+
+FieldReferenceGraph::BasicBlockHelperInfo *
+FieldReferenceGraph::getBasicBlockHelperInfo(const BasicBlock *BB) const {
+  auto it = BBInfoMap.find(BB);
+  assert(it != BBInfoMap.end());
+  return it->second;
+}
+
+FieldReferenceGraph::Node *
+FieldReferenceGraph::createNewNode(FieldNumType FieldNum, DataBytesType S) {
```

Comment:
Thanks! I was also thinking the name is weird. I'm gonna call it DistanceInBytesType